src/v/pandaproxy/schema_registry/sharded_store.cc

```diff
@@ -178,19 +158,55 @@ sharded_store::project_ids(subject_schema schema) {
             s_id = schema.id;
             vlog(plog.debug, "project_ids: using supplied ID {}", s_id.value());
         }
-    } else if (!s_id) {
+    } else if (s_id) {
+        vlog(plog.debug, "project_ids: existing ID {}", s_id.value());
+    }
+
+    // Determine if the subject already has a version that references this
+    // schema, deleted versions are seen.
+    const auto& sub = schema.schema.sub();
+    const auto versions = co_await _store.invoke_on(
+      shard_for(sub),
+      _smp_opts,
+      [sub](auto& s) -> std::vector<subject_version_entry> {
+          auto res = s.get_version_ids(sub, include_deleted::no);
+          if (
+            res.has_error()
+            && res.assume_error().code() == error_code::subject_not_found) {
+              return {};
+          }
+          return res.value();
+      });
+
+    auto has_version = s_id.has_value()
+                       && absl::c_any_of(
+                         versions, [id = *s_id](auto const& s_id_v) {
+                             return s_id_v.id == id;
+                         });
+
+    // Check compatibility of the schema
+    if (!has_version && !versions.empty()) {
+        auto compat = co_await is_compatible(
+          versions.back().version, schema.schema);
+        if (!compat) {
+            throw exception(
+              error_code::schema_incompatible,
+              fmt::format(
+                "Schema being registered is incompatible with an earlier "
+                "schema for subject \"{}\"",
+                sub));
+        }
+    }
```

Comment:
nit: should we log something here that indicates the schema is backwards compatible? I'm thinking about the case where a `post_subjects_subject_versions` fails and we don't know why.

Reply:
Nice. I'll ensure it's logged somewhere. Kinda makes sense to log all failures, so I might put it somewhere more general.